devito/core/operator.py

```diff
@@ -67,6 +69,11 @@ class BasicOperator(Operator):
     intensity of the generated kernel.
     """
 
+    SCALAR_MIN_TYPE = np.float16
```

Comment:
I think, instead of setting a specific dtype, we should have keywords such as "default" or "promote" or "downgrade", which the compiler interprets as expected e.g. if "promote" and `f.dtype==np.float16`, it will use `np.float32` for the temps

also, since it's a mixed precision feature, I'd call it something along the lines of  `MIXED_PREC_TEMP_TYPE` (in the future we might have additional MIXED_PREC_ attributes)